### DISCUSSION TITLE
Github secret truncated after "$"

### DISCUSSION THREAD

**Discussion by Amjad50** [community/community#25651] [Actions]
<p>Hello,</p>
<p>I found a bug in github secrets.<br>
If the secret contains “$”, for example “hello$world”. Then when used in github actions, we will only get “hello” and everything after and including the dollar sign will be removed.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>How are you using the secret in your workflow? That behavior sounds like what would happen if you put it into a shell command unquoted, and the shell interpreted <code>$world</code> as reading a variable.</p>

**Reply by danielo515** [reply]
What is the proper way? If I quote it using single quotes it seems to be evaluated (according to the debug log) to `''whatever''`, note the double quotes, which I don't think it's correct. However, using any other quotes, or no quotes does not work at all

**Reply by Amjad50** [reply]
From what I understand, github actions will replace the secret variables with the string u provided and that's it, so if you have a command lets say `echo ${{ secrets.SuperSecret }}`, then if you have the string "hello$world" in your secret. the whole expression will be `echo hello$world`, then when running `$world` would become empty string, which would result in printing `hello` only.

You should use single quotes in your command as such `echo '${{ secrets.SuperSecret }}'`,  which will become `echo 'hello$world'` and work normally.

What are you using so that you got 2 quotes?

**Reply by airtower-luna** [reply]
The proper way is usually to put the secret into an environment variable *for that step*.

```yaml
steps:
  - run: some-command "${MY_SECRET}"
    env:
      MY_SECRET: ${{ secrets.my_secret }}
```

The reason to do it this way: Actions syntax (the `${{ ... }} thing) is *literally* substituted before the shell ever sees your commands. So if there'd be, say, a single quote in your secret things would break again. With the quoted environment variable the shell is aware that there's a variable can treat it properly. :slightly_smiling_face:

**Reply by danielo515** [reply]
> What are you using so that you got 2 quotes?

I can't say for sure. I was just checking the debug log of the action and I saw that it was showing `''**''` in the log, so I took for granted it was like that.

**Reply by danielo515** [reply]
> The proper way is usually to put the secret into an environment variable _for that step_.
> 
> ```yaml
> steps:
>   - run: some-command "${MY_SECRET}"
>     env:
>       MY_SECRET: ${{ secrets.my_secret }}
> ```
> 
> The reason to do it this way: Actions syntax (the `${{ ... }} thing) is _literally_ substituted before the shell ever sees your commands. So if there'd be, say, a single quote in your secret things would break again. With the quoted environment variable the shell is aware that there's a variable can treat it properly. 🙂

That is an awesome answer. Thank you very much, will take note.

**Comment by Amjad50** [comment]
<p>That is correct, Thank you</p>
